subsys/usb/usb_c/usbc_pe_src_states.c

```diff
@@ -0,0 +1,755 @@
+			pe_set_state(dev, PE_SRC_DISCOVERY);
+		}
+		/*
+		 * Transition to the PE_SRC_Disabled state when:
+		 *	1) The Port Partners have not been PD Connected
+		 *	2) And the NoResponseTimer times out
+		 *	3) And the HardResetCounter > nHardResetCount
+		 */
+		else if ((atomic_test_bit(&pe->flags, PE_FLAGS_HAS_BEEN_PD_CONNECTED) == false)
+				&& usbc_timer_expired(&pe->pd_t_no_response)
+				&& pe->hard_reset_counter > PD_N_HARD_RESET_COUNT) {
+			pe_set_state(dev, PE_SUSPEND);
+		}
+		/*
+		 * The Policy Engine Shall go to the ErrorRecovery state when:
+		 *	1) The Port Partners have previously been PD Connected
+		 *	2) And the NoResponseTimer times out
+		 *	3) And the HardResetCounter > nHardResetCount
+		 */
+		else if (atomic_test_bit(&pe->flags, PE_FLAGS_HAS_BEEN_PD_CONNECTED)
+				&& usbc_timer_expired(&pe->pd_t_no_response)
+				&& pe->hard_reset_counter > PD_N_HARD_RESET_COUNT) {
+			usbc_request(dev, REQUEST_TC_ERROR_RECOVERY);
+		}
+
+		break;
+	case 1:
+		/*
+		 * The Policy Engine Shall transition to the PE_SRC_Negotiate_Capability state when:
+		 *	1) A Request Message is received from the Sink.
+		 */
+		if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_MSG_RECEIVED)) {
+			union pd_header header = prl_rx->emsg.header;
+
+			if ((header.extended == false) &&
+				received_data_message(dev, header, PD_DATA_REQUEST)) {
+				/* Set to highest revision supported by both ports */
+				prl_set_rev(dev, PD_PACKET_SOP,
+					MIN(PD_REV30, header.specification_revision));
+				pe_set_state(dev, PE_SRC_NEGOTIATE_CAPABILITY);
+			}
+		}
+		/*
+		 * The Policy Engine Shall transition to the PE_SRC_Hard_Reset
+		 * state when:
+		 *	1) The SenderResponseTimer times out
+		 */
+		else if (usbc_timer_expired(&pe->pd_t_sender_response)) {
+			pe_set_state(dev, PE_SRC_HARD_RESET);
+		}
+		break;
+	}
+}
+
+void pe_src_send_capabilities_exit(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+
+	usbc_timer_start(&pe->pd_t_sender_response);
+}
+
+/**
+ * @brief 8.3.3.2.4 PE_SRC_Negotiate_Capability State
+ */
+void pe_src_negotiate_capability_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+
+	/* Get sink request */
+	pe->snk_request = *(uint32_t *)prl_rx->emsg.data;
+
+	/*
+	 * Ask the Device Policy Manager to evaluate the Request
+	 * from the Attached Sink.
+	 */
+	pe->snk_request_reply =
+		policy_check_sink_request(dev, pe->snk_request);
+
+	LOG_INF("PE_SRC_Negotiate_Capability");
+}
+
+void pe_src_negotiate_capability_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * The Policy Engine Shall transition to the
+	 * PE_SRC_Transition_Supply state when:
+	 *	1) The Request can be met.
+	 */
+	if (pe->snk_request_reply == SNK_REQUEST_VALID) {
```

Comment:
Could the run state here just be collapsed into the entry state?

Reply:
Done